docs/design/specs/Memory-model.md

```diff
@@ -74,14 +74,16 @@ The effects of ordinary reads and writes can be reordered as long as that preser
      - IL load instructions with `volatile.` prefix when instruction supports such prefix
      - `System.Threading.Volatile.Read`
      - `System.Thread.VolatileRead`
-     - Acquiring a lock (`System.Threading.Monitor.Enter` or entering a synchronized method)
+     - `System.Threading.Volatile.ReadBarrier` (applies to all prior reads)
+     - Acquiring a lock (`System.Threading.Monitor.Enter` or entering a synchronized method, applies to all prior reads)
 
 * **Volatile writes** have "release semantics" - the effects of a volatile write will not be observable before effects of all previous, in program order, reads and writes become observable.
   Operations with release semantics:
      - IL store instructions with `volatile.` prefix when such prefix is supported
      - `System.Threading.Volatile.Write`
      - `System.Thread.VolatileWrite`
-     - Releasing a lock (`System.Threading.Monitor.Exit` or leaving a synchronized method)
+     - `System.Threading.Volatile.WriteBarrier` (applies to all future writes)
```

Comment:
```suggestion
     - `System.Threading.Volatile.WriteBarrier` (applies to all following writes)
```

Maybe less ambiguous as we refer to the program order here vs. order of execution effects.